### PR TITLE
[compatibility] Make struct removal violate layout compatibility

### PR DESCRIPTION
**Pull request by wrwg**
## Motivation

Until now, when a struct was removed, this only counted as a `struct_and_function_linkage` compatibility issue. However, it is also a `struct_layout` issue because it opens the door to first remove a struct, then re-add it with a different layout.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes

## Test Plan

Existing

### CODE REVIEW

**Review by sblackshear — APPROVED** [review]
Very good catch!

We need to come up with a more direct way to test these. But in the meantime, would you mind adding an expected value test for this scenario (https://github.com/move-language/move/tree/main/language/tools/move-cli/tests/sandbox_tests/republish is a helpful example)?

I recall that there use to be a lot more expected value tests for the compatibility logic, I wonder if some of them were part of the `diem-cli` testsuite and got deleted...

**Comment by wrwg** [comment]
> Very good catch!

Credits go to @chen-robert from OtterSec.

> 
> We need to come up with a more direct way to test these. But in the meantime, would you mind adding an expected value test for this scenario (https://github.com/move-language/move/tree/main/language/tools/move-cli/tests/sandbox_tests/republish is a helpful example)?
> 
> I recall that there use to be a lot more expected value tests for the compatibility logic, I wonder if some of them were part of the `diem-cli` testsuite and got deleted...

Yes those tests are (to some extent) in the Aptos repo.

However, I saw @jolestar has been writing independently some tests. If I land this could you afterwards add the tests, @jolestar? Transactional tests seem the right thing to do here. (Not acquainted myself with that infra.)

**Review by chen-robert — COMMENTED** [review]
looks good to me -- would also note that there are probably many other examples of this issue in these compatibility checks, will need some more time to review on our side.

**Review by jolestar — APPROVED** [review]

**Comment by jolestar** [comment]
> However, I saw @jolestar has been writing independently some tests. If I land this could you afterwards add the tests, @jolestar? Transactional tests seem the right thing to do here. (Not acquainted myself with that infra.)

Yes. I will continue the #500. This PR land should break the test case https://github.com/move-language/move/pull/500/files#diff-78d81d81673c1eeafe39de4b8007dae712660f9e869e24e8c8721c0f8ecd3c31 .

File: language/move-binary-format/src/compatibility.rs [hunk 1]
```diff
@@ -50,17 +51,10 @@ impl Compatibility {
...
-                    // this struct `Parent::T`. That is, code like the function `foo` in
-                    // ```
-                    // struct S { t: Parent::T }
-                    // public fun foo(a: addr): S { move_from<S>(addr) }
-                    // ```
-                    // in module `Child` will continue to run without error. But values of type
-                    // `Parent::T` in `Child` are now "orphaned" in the sense that `Parent` no
-                    // longer exposes any API for reading/writing them.
+                    struct_layout = false;
                     continue;
```
Comment (zsluedem) [inline language/move-binary-format/src/compatibility.rs:58]: Just a nitpick. It could `break` or even return the result directly here.

**Pull request merged**